PWGHF: Modified Task Dplus adding THnSparse for ML

Luca610 opened the pull request:
This PR introduces two process functions to the Dplus Task to fill THnSparse objects with candidate Mass, Pt and BDT scores for ML analysis .

The task's architecture now includes template functions for histogram filling, aiming to simplify future modifications and expansions.

fcatalan92 commented on PWGHF/D2H/Tasks/taskDplus.cxx:
```diff
@@ -94,127 +112,217 @@ struct HfTaskDplus {
-      if (!(candidate.hfflag() & 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
-        continue;
-      }
+    // not possible in Filter since expressions do not support binary operators
+    if (!(candidate.hfflag() & 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
+      return;
+    }
```
It is now possible to do this. You could add a filter taking https://github.com/AliceO2Group/O2Physics/blob/3301cc27ae21395c526e867ce7ef6ee98ed41d84/PWGHF/D2H/Tasks/taskDs.cxx#L84 as example.

fcatalan92 commented on PWGHF/D2H/Tasks/taskDplus.cxx:
```diff
@@ -94,127 +112,217 @@ struct HfTaskDplus {
+      registry.fill(HIST("hd0Prong2"), candidate.impactParameter2(), pt);
+  }
+
+  // Fill THnSparses for the reconstructed Dplus candidates ML analysis for the reconstructed Dplus candidates
```
```suggestion
  // Fill THnSparses for the ML analysis
```

fcatalan92 commented on PWGHF/D2H/Tasks/taskDplus.cxx:
```diff
@@ -94,127 +112,217 @@ struct HfTaskDplus {
+  template <bool isMc, typename T1>
+  void fillSparseML(const T1& candidate)
+  {
+    if (!(candidate.hfflag() & 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi) || (yCandRecoMax >= 0. && std::abs(hfHelper.yDplus(candidate)) > yCandRecoMax)) {
+      return;
+    }
```
Same comment as above for the `hfflag` check

fcatalan92 commented on PWGHF/D2H/Tasks/taskDplus.cxx:
```diff
@@ -94,127 +112,217 @@ struct HfTaskDplus {
-      // not possible in Filter since expressions do not support binary operators
-      if (!(candidate.hfflag() & 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
-        continue;
+    // not possible in Filter since expressions do not support binary operators
+    if (!(candidate.hfflag() & 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi)) {
+      return;
+    }
```
Same comment as above

fcatalan92 commented on PWGHF/D2H/Tasks/taskDplus.cxx:
```diff
@@ -94,127 +112,217 @@ struct HfTaskDplus {
+    if (yCandRecoMax >= 0. && std::abs(hfHelper.yDplus(candidate)) > yCandRecoMax) {
+      return;
+    }
+    if (std::abs(candidate.flagMcMatchRec()) == 1 << aod::hf_cand_3prong::DecayType::DplusToPiKPi) {
```
Also this can be moved to a Filter or Partition

fcatalan92 reviewed (changes requested):
Hi @Luca610, thanks for the developments! I added some comments, mainly on the possibility to better use Filters and Partitions (the task was written some time ago and the capabilities of the framework were different). Another general comment is on the `fillHisto` functions: I would keep in there only the histogram filling and move all the other logic outside. In this way you can avoid something like this `fillHistoMCRec(const T1& candidate, const T2& mcParticles)` which looks odd.

Luca610 commented:
Hi @fcatalan92, thank you for your comments! 
I implemented them in a new version of the task. I added some runAnalysis functions to remove logic checks from the fillHisto functions and added the filter on the hfflag.
Let me know if you have other comments!

Thanks again!

fcatalan92 reviewed (changes requested):
Hi @Luca610, thanks for implementing my suggestions. I have still a small comment.

Luca610 commented on PWGHF/D2H/Tasks/taskDplus.cxx:
```diff
@@ -94,127 +124,237 @@ struct HfTaskDplus {
+        if ((yCandRecoMax >= 0. && std::abs(hfHelper.yDplus(candidate)) > yCandRecoMax)) {
+          continue;
+        }
+        fillSparseML<false, false>(candidate);
```
You're right my mistake!
